FormatCommand verify does not print a clear error message

Found this running on CI:

```
Run ./<sidekick-cli> format --verify
Verifying package:foo
Formatted 7 files (0 changed) in 0.32 seconds.
Verifying package:bar
Changed /__w/bar/.../x.dart
Changed /__w/bar/.../y.dart
Formatted 17 files (2 changed) in 0.46 seconds.
Unhandled exception:
Instance of 'DartFileFormatException'
#0      FormatCommand._verifyThrow (package:sidekick_core/src/commands/format_command.dart:132)
#1      FormatCommand.run (package:sidekick_core/src/commands/format_command.dart:126)
#2      CommandRunner.runCommand (package:args/command_runner.dart:212)
#3      SidekickCommandRunner.run (package:sidekick_core/sidekick_core.dart:291)
Error: Process completed with exit code 255.
```

We should print a clear error message like `Dart files are not correctly formatted. Run "${SidekickContext.cliName} format" to format the code.` (in red)

--- a/sidekick_core/lib/src/commands/format_command.dart
+++ b/sidekick_core/lib/src/commands/format_command.dart
@@ -65,6 +65,7 @@ class FormatCommand extends Command {
   }
 
   bool foundFormatError = false;
+  List<String> unformattedFiles = [];
 
   @override
   Future<void> run() async {
@@ -129,10 +130,11 @@ class FormatCommand extends Command {
   void _verifyThrow() {
     final bool verify = argResults?['verify'] as bool? ?? false;
     if (verify && foundFormatError) {
-      throw DartFileFormatException(
-        'Dart files are not correctly formatted. '
-        'Run "${SidekickContext.cliName} format" to format the code.',
-      );
+      final message = 'Following Dart files are not formatted correctly:\n'
+          '${unformattedFiles.join('\n')}\n'
+          'Run "${SidekickContext.cliName} format" to format the code.';
+      printerr(red(message));
+      throw DartFileFormatException(message);
     }
   }
 
@@ -151,6 +153,8 @@ class FormatCommand extends Command {
       print("No files to format");
       return;
     }
+    final progress =
+        verify ? Progress.capture() : Progress.print(capture: true);
     final exitCode = dart(
       [
         'format',
@@ -164,10 +168,15 @@ class FormatCommand extends Command {
       nothrow: verify,
       // Lines like `Changed x.dart`, `Formatted x files (y changed) in z seconds`
       // should only be printed when the change is actually written to the files (when verify is false)
-      progress: verify ? Progress.devNull() : Progress.print(),
+      progress: progress,
     );
     if (exitCode != 0) {
       foundFormatError = true;
+      unformattedFiles.addAll(
+        progress.lines
+            .where((it) => it.startsWith('Changed '))
+            .map((it) => it.substring('Changed '.length)),
+      );
     }
   }
 }
@@ -232,5 +241,5 @@ class DartFileFormatException implements Exception {
   DartFileFormatException(this.message);
 
   @override
-  String toString() => message;
+  String toString() => 'DartFileFormatException{message: $message}';
 }

--- a/sidekick_core/lib/src/commands/format_command.dart
+++ b/sidekick_core/lib/src/commands/format_command.dart
@@ -131,7 +131,7 @@ class FormatCommand extends Command {
     if (verify && foundFormatError) {
       throw DartFileFormatException(
         'Dart files are not correctly formatted. '
-        'Run "${SidekickContext.cliName} sidekick format" to format the code.',
+        'Run "${SidekickContext.cliName} format" to format the code.',
       );
     }
   }
@@ -227,4 +227,7 @@ class DartFileFormatException implements Exception {
   final String message;
 
   DartFileFormatException(this.message);
+
+  @override
+  String toString() => message;
 }

--- a/sidekick_core/lib/src/commands/format_command.dart
+++ b/sidekick_core/lib/src/commands/format_command.dart
@@ -162,6 +162,9 @@ class FormatCommand extends Command {
         if (verify) '--output=none',
       ],
       nothrow: verify,
+      // Lines like `Changed x.dart`, `Formatted x files (y changed) in z seconds`
+      // should only be printed when the change is actually written to the files (when verify is false)
+      progress: verify ? Progress.devNull() : Progress.print(),
     );
     if (exitCode != 0) {
       foundFormatError = true;

--- a/sidekick_core/test/format_command_test.dart
+++ b/sidekick_core/test/format_command_test.dart
@@ -318,7 +318,8 @@ name: dashi
 
     test('--verify throws, but does not format', () async {
       await insideFakeProjectWithSidekick((dir) async {
-        dir.file('some.dart').writeAsStringSync(_dartFile140);
+        final unformattedFile = dir.file('some.dart')
+          ..writeAsStringSync(_dartFile140);
         final fakeStdout = FakeStdoutStream();
         final fakeStderr = FakeStdoutStream();
         overrideIoStreams(
@@ -331,20 +332,25 @@ name: dashi
             runner.addCommand(FormatCommand());
             expectLater(
               () => runner.run(['format', '--verify']),
-              throwsA(
-                isA<DartFileFormatException>().having(
-                  (p0) => p0.toString(),
-                  'String representation',
-                  'Dart files are not correctly formatted. Run "dash format" to format the code.',
-                ),
-              ),
+              throwsA(isA<DartFileFormatException>()),
             );
 
             expect(dir.file('some.dart').readAsStringSync(), _dartFile140);
           },
         );
 
-        expect(fakeStderr.lines, isEmpty);
+        expect(
+          fakeStderr.lines,
+          contains(
+            stringContainsInOrder(
+              [
+                'Following Dart files are not formatted correctly:',
+                unformattedFile.path,
+                'Run "dash format" to format the code.'
+              ],
+            ),
+          ),
+        );
         // shouldn't print lines like `Changed x.dart`, `Formatted x files (y changed) in z seconds`
         expect(
           fakeStdout.lines,

--- a/sidekick_core/test/format_command_test.dart
+++ b/sidekick_core/test/format_command_test.dart
@@ -324,7 +324,13 @@ name: dashi
         runner.addCommand(FormatCommand());
         expectLater(
           () => runner.run(['format', '--verify']),
-          throwsA(isA<DartFileFormatException>()),
+          throwsA(
+            isA<DartFileFormatException>().having(
+              (p0) => p0.toString(),
+              'String representation',
+              'Dart files are not correctly formatted. Run "dash format" to format the code.',
+            ),
+          ),
         );
 
         expect(dir.file('some.dart').readAsStringSync(), _dartFile140);

--- a/sidekick_core/test/format_command_test.dart
+++ b/sidekick_core/test/format_command_test.dart
@@ -1,5 +1,6 @@
-import 'package:sidekick_core/sidekick_core.dart';
+import 'package:sidekick_core/sidekick_core.dart' hide isEmpty;
 import 'package:sidekick_core/src/commands/format_command.dart';
+import 'package:sidekick_test/fake_stdio.dart';
 import 'package:sidekick_test/sidekick_test.dart';
 import 'package:test/test.dart';
 
@@ -318,22 +319,75 @@ name: dashi
     test('--verify throws, but does not format', () async {
       await insideFakeProjectWithSidekick((dir) async {
         dir.file('some.dart').writeAsStringSync(_dartFile140);
-        final runner = initializeSidekick(
-          dartSdkPath: systemDartSdkPath(),
-        );
-        runner.addCommand(FormatCommand());
-        expectLater(
-          () => runner.run(['format', '--verify']),
-          throwsA(
-            isA<DartFileFormatException>().having(
-              (p0) => p0.toString(),
-              'String representation',
-              'Dart files are not correctly formatted. Run "dash format" to format the code.',
-            ),
-          ),
+        final fakeStdout = FakeStdoutStream();
+        final fakeStderr = FakeStdoutStream();
+        overrideIoStreams(
+          stderr: () => fakeStderr,
+          stdout: () => fakeStdout,
+          body: () {
+            final runner = initializeSidekick(
+              dartSdkPath: systemDartSdkPath(),
+            );
+            runner.addCommand(FormatCommand());
+            expectLater(
+              () => runner.run(['format', '--verify']),
+              throwsA(
+                isA<DartFileFormatException>().having(
+                  (p0) => p0.toString(),
+                  'String representation',
+                  'Dart files are not correctly formatted. Run "dash format" to format the code.',
+                ),
+              ),
+            );
+
+            expect(dir.file('some.dart').readAsStringSync(), _dartFile140);
+          },
         );
 
-        expect(dir.file('some.dart').readAsStringSync(), _dartFile140);
+        expect(fakeStderr.lines, isEmpty);
+        // shouldn't print lines like `Changed x.dart`, `Formatted x files (y changed) in z seconds`
+        expect(
+          fakeStdout.lines,
+          ['Verifying package:dash', 'Verifying package:main_project'],
+        );
+      });
+    });
+
+    test('prints informative output when formatting files', () async {
+      await insideFakeProjectWithSidekick((dir) async {
+        final fakeStdout = FakeStdoutStream();
+        final fakeStderr = FakeStdoutStream();
+        await overrideIoStreams(
+          stderr: () => fakeStderr,
+          stdout: () => fakeStdout,
+          body: () async {
+            setupProject(dir, mainContent: _dartFile80);
+            final runner = initializeSidekick(
+              dartSdkPath: systemDartSdkPath(),
+            );
+            runner.addCommand(FormatCommand());
+            await runner.run(['format']);
+          },
+        );
+
+        expect(fakeStderr.lines, isEmpty);
+        expect(
+          fakeStdout.lines,
+          containsAll(
+            [
+              'Formatting package:dash',
+              stringContainsInOrder(
+                ['Formatted', 'packages/dash/lib/dash_sidekick.dart'],
+              ),
+              stringContainsInOrder(
+                ['Formatted', 'packages/dash/lib/src/dash_project.dart'],
+              ),
+              stringContainsInOrder(
+                ['Formatted 2 files (2 changed) in', 'seconds'],
+              ),
+            ],
+          ),
+        );
       });
     });
   });

